### PR TITLE
Make the k8s client rate limiter configurable

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

Make the following k8s client rate limiter configurable:

 * virt-api: default rate limiter used in e.g. the vnc or console resources
 * virt-api: webhook rate limiter with very high values which is responsible for validations
 * virt-controller: rate limiter for our controllers (now configured to the same values like the k8s controller-manager)
 * virt-handler: rate limiter for our node daemon (now configured to the same defaults like the kubelet)
 
 The values can be changed via our `KubeVirt` CR and are picked up without restart by our components.

The default configuration looks like this:

```yaml
configuration:
  apiConfiguration:
    restClient:
      rateLimiter:
        tokenBucketRateLimiter:
          burst: 10
          qps: 5
  controllerConfiguration:
    restClient:
      rateLimiter:
        tokenBucketRateLimiter:
          burst: 30
          qps: 20
  handlerConfiguration:
    restClient:
      rateLimiter:
        tokenBucketRateLimiter:
          burst: 10
          qps: 5
  webhookConfiguration:
    restClient:
      rateLimiter:
        tokenBucketRateLimiter:
          burst: 400
          qps: 200
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Make k8s client rate limits configurable
```

### CODE REVIEW

File: pkg/util/ratelimiter/ratelimiter.go
Comment (davidvossel) [inline pkg/util/ratelimiter/ratelimiter.go:45]: it probably won't make a huge difference, but can we get away with a RWMutex? https://pkg.go.dev/sync#RWMutex

**Review by davidvossel — COMMENTED** [review]
makes sense

File: pkg/util/ratelimiter/ratelimiter.go
Comment (rmohr) [reply pkg/util/ratelimiter/ratelimiter.go:45]: I am not sure if a rw mutex is faster since I don't know how it is implemented in go (the gating and reference counting can have more overhead than just looking up the reference).

Comment (rmohr) [reply pkg/util/ratelimiter/ratelimiter.go:45]: I will check the performance characteristics.

Comment (marceloamaral) [reply pkg/util/ratelimiter/ratelimiter.go:45]: This shouldn't be called too much time, shouldn't it? Only when updating or starting the API. Or, Am I missing something? 
If it is not being called too much time, regular Lock and Unlock should be fine here.

**Review by marceloamaral — APPROVED** [review]
It makes sense and looks good to me

File: pkg/util/ratelimiter/ratelimiter.go
Comment (rmohr) [reply pkg/util/ratelimiter/ratelimiter.go:45]: I could not find a reproducable benefit for any of the following methods: `RWMutext with RLock` vs. `Mutex with Lock` vs. `always Lock with RWMutex`. 

I will leave it as is.

Comment (marceloamaral) [reply pkg/util/ratelimiter/ratelimiter.go:45]: It is up to @davidvossel, the current approach seems ok to me since you did not experienced any performance difference.

**Review by davidvossel — COMMENTED** [review]
/approve
/hold

I put the hold on here to discuss if there's any practical way we can test that the rate limiter config is picked up and performing as expected.

I'm thinking about this primarily from a defensive angle to protect the feature. Your logic looks like it works and I know you've tested it locally.  Is there anything we can do to add some sort of test to validate it keeps working as expected so we don't run into a surprise later on?

**Comment by marceloamaral** [comment]
Regarding the test, I have an idea of increasing and decreasing the QPS.

1.  Make the rate limit very small and verify the Rest Rate Limiter Duration
`histogram_quantile(0.95, sum(rate(rest_client_rate_limiter_duration_seconds_bucket{}[5m])) by (container, le, url))`
We should see high values. We could also verify one module only, e.g., `container="virt-controller"`

2. Make the rate limit very high and verify the Rest Rate Limiter Duration. We should see very low values.

**Comment by rmohr** [comment]
> Regarding the test, I have an idea of increasing and decreasing the QPS.
> 
> 1. Make the rate limit very small and verify the Rest Rate Limiter Duration
>    `histogram_quantile(0.95, sum(rate(rest_client_rate_limiter_duration_seconds_bucket{}[5m])) by (container, le, url))`
>    We should see high values. We could also verify one module only, e.g., `container="virt-controller"`
> 2. Make the rate limit very high and verify the Rest Rate Limiter Duration. We should see very low values.

Sounds good. I will add something along that line.

**Comment by rmohr** [comment]
@davidvossel @marceloamaral too a more primitive route regarding to testing by simply checking if the VMI creations and starts take significantly longer. Let me know what you thinkg. PTAL.

**Comment by rmohr** [comment]
/unhold

PTAL

**Comment by davidvossel** [comment]
These thresholds for the tests might be cutting it close. saw this in flake finder

```
13:46:32: • Failure [78.713 seconds]
13:46:32: [Serial][sig-compute]Infrastructure changes to the kubernetes client [It] on the virt handler rate limiter should lead to delayed VMI running states 
13:46:32: tests/infra_test.go:138
13:46:32: 
13:46:32:   Expected
13:46:32:       <float64>: 33.461093293
13:46:32:   to be >
13:46:32:       <float64>: 34.946444026500004
```

**Review by davidvossel — APPROVED** [review]
/lgtm

you may want to loosen some of the test thresholds related to timing if tests prove flaky

**Comment by rmohr** [comment]
/hold

Tests are too flaky

File: tests/infra_test.go
Comment (marceloamaral) [inline tests/infra_test.go:143]: Why do you need to force the placement decision?

File: tests/infra_test.go
Comment (marceloamaral) [inline tests/infra_test.go:124]: Why you set Bust=3 and QPS=2 here, and in the handler test Bust=1 and QPS=1?
Can this test also use Bust=1 and QPS=1?

File: tests/infra_test.go
Comment (marceloamaral) [inline tests/infra_test.go:175]: Is there any reason why you expect the rate limit make the latency 50% slower?
Instead of relative times, it might be better to have absolute times, as used in the first test.

Comment (marceloamaral) [reply tests/infra_test.go:175]: Maybe this is where the flaky test that David pointed before is.
The first test `Expect(slowDuration.Seconds()).To(BeNumerically(">", 5*fastDuration.Seconds()))` you used the relative threshold as 5 times higher than the fastDuration; the second test is 1.5 times.
If you prefer to use relative threshold instead of absolute ones, maybe both tests should have similar thresholds?

**Review by marceloamaral — CHANGES_REQUESTED** [review]

File: tests/infra_test.go
Comment (rmohr) [reply tests/infra_test.go:143]: Otherwise the VMs get distributed to differend virt-handlers which will make the rate-limit throttling less predictable.

File: tests/infra_test.go
Comment (rmohr) [reply tests/infra_test.go:124]: If I set it too low we have the issue that the leader-election will not work either and we have virt-controllers crash-looping.

File: tests/infra_test.go
Comment (rmohr) [reply tests/infra_test.go:175]: I have to make it relative. Otherwise the tests could never succeed locally and in CI. I choose here only 1.5 since the difference from create-to-running is from a factor perspective less big when throttling only scheduled-to-running, vs created-to-scheduled above.

Comment (marceloamaral) [reply tests/infra_test.go:194]: I saw that you have updated how the `fastDuration` is measured.
Looks good!

File: tests/infra_test.go
Comment (rthallisey) [inline tests/infra_test.go:155]: This kind of test I was thinking we should leave to be measured by the perf/scale framework.  QPS improves performance by reducing the delta between when a virt-launcher pod is Running and when a VMI is Running, which is perfect for a threshold test using the new framework. 

For a func test to validate this feature, I was thinking a test that checks the number of times we get rate limited with low and high qps.  That will tell us if our default values ever get too low again or if they are too high because we made a fix that lowers the number of API calls we make.

Comment (rmohr) [reply tests/infra_test.go:155]: > This kind of test I was thinking we should leave to be measured by the perf/scale framework. QPS improves performance by reducing the delta between when a virt-launcher pod is Running and when a VMI is Running, which is perfect for a threshold test using the new framework.

Definitely. This test does not imply that that these values are good or bad and they are not subject of change. They should simply state that the rate limiter settings have an effect which makes things slower.

Comment (rthallisey) [reply tests/infra_test.go:155]: Right, but the most accurate thing to measure here is the number of times we hit the rate limiter based on the QPS.  Slowness is a symptom that can be affected by many things, so I want to leave that measurement to the perf framework

Comment (rmohr) [reply tests/infra_test.go:155]: > Slowness is a symptom that can be affected by many things

Yes, that is why this takes a baseline first.

I understand what you want to get at, still I want to ask for doing this with reasonable complexity. It is just pretty simple and should be pretty robust this way, while going to the metric and trying to interpret it is more complex.

File: tests/infra_test.go
Comment (marceloamaral) [reply tests/infra_test.go:141]: In this case, can we use Bust=3 and QPS=2 for all small cases?
We could also include a comment that this is the minimal requirement to make sure that leader-election will work.
So, anyone can understand that in the future.

File: tests/infra_test.go
Comment (marceloamaral) [reply tests/infra_test.go:155]: Since the measured latencies are relative, the test does not characterize performance.

Although the @rthallisey idea of counting the rate limited calls is very interesting, it requires collect prometheus metrics which complicates the test. First, prometheus is not deployed by default in the functional tests. Second, it requires more coding...

**Comment by marceloamaral** [comment]
The tests are still flaky.
```
Tests Suite: [Serial][sig-compute]Infrastructure changes to the kubernetes client on the controller rate limiter should lead to delayed VMI starts 19s
tests/infra_test.go:125 Expected     <float64>: 7.042141059 to be >     <float64>: 10.145289255000002 tests/infra_test.go:152
```

The problem is in:
`Expect(slowDuration.Seconds()).To(BeNumerically(">", 5*fastDuration.Seconds()))`

So, the fastDuration is ~2s and the slowDuration is ~3x slower not 5x.

I am afraid if these latencies have a big fluctuation the test will always be flaky.

Therefore, maybe, you could do something similar to the  @rthallisey idea of counting the rate limited calls, but without requiring Prometheus. For example, what about read the logs?
https://github.com/kubernetes/client-go/blob/c8b4c1e4bd59fd6dab2a5c8b1600bf369892a5f4/rest/request.go#L602
If the throttle latency is higher than `longThrottleLatency` and/or `extraLongThrottleLatency` it will appear in the log, so we can count them.

**Comment by fabiand** [comment]
DO we have a chance to get this into the next release?

**Comment by rthallisey** [comment]
>Release note:

>Make k8s client rate limits configurable
>

If we can measure it, it would be good to mention the perf impact

**Comment by rmohr** [comment]
> The tests are still flaky.
> 
> ```
> Tests Suite: [Serial][sig-compute]Infrastructure changes to the kubernetes client on the controller rate limiter should lead to delayed VMI starts 19s
> tests/infra_test.go:125 Expected     <float64>: 7.042141059 to be >     <float64>: 10.145289255000002 tests/infra_test.go:152
> ```
> 
> The problem is in:
> `Expect(slowDuration.Seconds()).To(BeNumerically(">", 5*fastDuration.Seconds()))`
> 
> So, the fastDuration is ~2s and the slowDuration is ~3x slower not 5x.
> 
> I am afraid if these latencies have a big fluctuation the test will always be flaky.
> 
> Therefore, maybe, you could do something similar to the @rthallisey idea of counting the rate limited calls, but without requiring Prometheus. For example, what about read the logs?
> https://github.com/kubernetes/client-go/blob/c8b4c1e4bd59fd6dab2a5c8b1600bf369892a5f4/rest/request.go#L602
> If the throttle latency is higher than `longThrottleLatency` and/or `extraLongThrottleLatency` it will appear in the log, so we can count them.

@rthallisey @marceloamaral good ideas. I will give it one last try with slight adjustmenst since I would prefer to keep it simple, but it continues to be flaky I will follow your advice.

**Comment by rmohr** [comment]
the flake check looks now good too. Triggering again to be sure.

/test pull-kubevirt-check-tests-for-flakes

**Comment by rmohr** [comment]
/Unhold
/retest
PTAL

**Comment by rmohr** [comment]
> @rmohr it seems that you need to merge with main to fix the conflicts

Done. @marceloamaral could you give it a lgtm if it looks good to you?

**Comment by marceloamaral** [comment]
Looks good! 
/lgtm

**Pull request merged**
